### PR TITLE
Feature/enable windows builds

### LINKED ISSUE #7: Plugin should work with Unix-style file path on Windows
When I specified

> <sourceDir>${basedir}/src/main/javascript</sourceDir>

It didn't work on Windows and always created resulted file in the source dir instead of the targetDir.

It happends because the following code from TargetFileWriter:

```java
        final String targetDirectory = sourceFile.toString()
                .replace(sourceFileName.toString(), "")
                .replace(sourceDir, targetDir);
```
tries to remove "C:\Users\slava\git\mystamps/src/main/javascript" from string "C:\Users\slava\git\mystamps\src\main\javascript\components\" and it couldn't do that.

As you can see, the slashes (/ vs \) are different.

It would be good if plugin will handle paths with any type of the directories separators similarly.

The current workaround is:
> <sourceDir>${basedir}\src\main\javascript</sourceDir>

but it means that the configuration can't work on Linux.

### PR DESCRIPTION
**Pull request by jonaskoperdraat**
Hi @milpol ,

First of all, thanks for writing this plugin! I'v been using it since a couple of weeks now, and I am glad that I can use modern day JavaScript in an application that requires the JavaScript to be less fancy, without the need to set up a full fledged node.js frontend build.

We recently ran into issue #7, and seeing as that issue was labeled 'help wanted', I figured I'd look into it and see if I could help out. This pull request does that. It fixes #7 by resorting to using `Path` as much as possible, and `File` where that is not possible. Also, when provisioning the `DirectoryScanner` with the includes and excludes, care is taken to convert Unix style path separators to the required separators. 

Incidentally I went a bit overboard with the refactoring. I took the liberty to restructure the plugin somewhat, to be able to provide better feedback to the user, and to reduce some of it's complexity. I hope you don't mind, and mostly I ask that you don't take me refactoring your code personally, as it is merely my attempt to contribute and make the plugin as simple and useful as possible.

Also, in the process of making the plugin platform independent, I have been thinking about #9 (performance), and I think I have pinpointed the bottleneck, and might have a solution, which I've written a unittest for  (src/test/java/com/jarslab/maven/babel/plugin/ReusableEngineTest.java). Apparently that also made it into this commit, which wasn't what I had intended, but hey 🤷‍♂ 

Seeing there has been quite some refactoring, I could understand if you're not willing to accept this PR. In that case no hard feelings 😉 

Kind regards,

Jonas Koperdraat

### CODE REVIEW

File: src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java [hunk 1]
```diff
@@ -63,100 +66,51 @@ public void execute() throws MojoExecutionException, MojoFailureException
...
-            if (parallel) {
-                getLog().info("Run in parallel mode.");
-                transpilers.parallel();
-            }
-            transpilers
-                    .peek(this::logExtractedFile)
-                    .forEach(BabelTranspiler::execute);
+            transpilations.parallelStream()
+                    .map(transpiler::execute)
+                    .forEach(targetFileWriter::writeTargetFile);
```
Comment (jonaskoperdraat) [inline src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java:102]: I see I've dropped the parallel processing mode here, which hasn't been quite intentional. I'l have to revert that. At least until I adopt reusing the `ScriptEngine` (`ReuseEngineTest#testReuseEngine`) in which case parallel transpiling is no longer an option (but it's much faster). 

- [x] Make parallel processing optional again

File: src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java [hunk 2]
```diff
@@ -13,70 +14,48 @@
...
+import static java.lang.String.format;
+
+@Builder
+class BabelTranspiler {
 
-class BabelTranspiler
-{
     private static final String JAVASCRIPT_MIME_TYPE = "text/javascript";
     private static final String INPUT_VARIABLE = "input";
     private static final String BABEL_EXECUTE = "Babel.transform(%s, {presets: [%s]}).code";
```
Comment (jonaskoperdraat) [inline src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java:25]: I'm thinking we could probably quite easily read the `option` parameter of `Babel.transform` (which is currently only `{presents: [...]}`) from a user-configurable file, allowing much greater flexibility.

File: src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java (hunk 1 above)
Comment (jonaskoperdraat) [reply src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java:102]: This has been addressed in #12

File: pom.xml [hunk 3]
```diff
@@ -70,6 +79,11 @@
             <version>2.21.0</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
+            <version>1.18.8</version>
+        </dependency>
```
Comment (milpol) [inline pom.xml:86]: Could you replace Lombok in favor of Immutables (https://immutables.github.io)? 
I know Lombok is kind of a standard now but personally I don't like it (have reasons, but probably this is not a place for such discussions).

File: src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java [hunk 4]
```diff
@@ -43,15 +47,14 @@
...
 
     @Override
-    public void execute() throws MojoExecutionException, MojoFailureException
-    {
+    public void execute() throws MojoExecutionException, MojoFailureException {
         final Charset charset = Charset.forName(encoding);
         if (verbose) {
             getLog().info("Run in the verbose mode.");
             getLog().info(String.format("Charset: %s.", charset));
+            getLog().debug(this.toString());
```
Comment (milpol) [inline src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java:55]: No need for `this.` reference?

File: src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java (hunk 1 above)
Comment (milpol) [reply src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java:102]: Makes sense (BTW I like the thread option idea), could we merge PR one after another? (I will postpone releasing new version until then).

File: src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java (hunk 2 above)
Comment (milpol) [reply src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java:25]: Good idea, I was thinking about that originally but frankly have no needs to implement it. Probably case for another PR?

File: src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java [hunk 5]
```diff
@@ -13,70 +14,48 @@
...
                     new FileInputStream(babelSource), charset);
             final ScriptEngine engine = new ScriptEngineManager(null)
                     .getEngineByMimeType(JAVASCRIPT_MIME_TYPE);
             final SimpleBindings simpleBindings = new SimpleBindings();
             engine.eval(fileReader, simpleBindings);
-            final String source = new String(Files.readAllBytes(sourceFilePath), charset);
-            if (verbose) {
-                log.debug(String.format("%s source:\n%s", sourceFilePath, source));
-            }
+            final String source = new String(Files.readAllBytes(context.getSource()), charset);
```
Comment (milpol) [inline src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java:45]: Minor thing, but shouldn't `.toString` do the thing?

File: src/test/java/com/jarslab/maven/babel/plugin/BabelTranspilerCp1252Test.java [hunk 6]
```diff
@@ -10,62 +10,59 @@
...
 import java.nio.file.Files;
 import java.nio.file.Path;
-import java.nio.file.Paths;
 
-import static org.mockito.ArgumentMatchers.argThat;
-import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.Mockito.times;
-import static org.mockito.Mockito.verify;
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (milpol) [inline src/test/java/com/jarslab/maven/babel/plugin/BabelTranspilerCp1252Test.java:18]: Hamcrest is a bit outdated library, could you rely on AssertJ?

**Review by milpol — COMMENTED** [review]
Great work, thank you.
My biggest concern is Lombok (which is somewhat childish probably, but I would really like to avoid it in my projects). 
Nevertheless, that's a lot of your work here, so if you don't want spend more time on throwing Lombok away we can merge it and I'll purge it on my own :)
Thanks again

**Comment by jonaskoperdraat** [comment]
No, I can get rid of Lombok, no sweat. Mind if I do that in #12? And care to elaborate on your desire to keep your projects Lombok-free?

**Review by milpol — APPROVED** [review]
Sure, no worries, I will merge this one so we could jump to #12.
Don't bother if you don't feel comfortable with `Immutables` or `AutoValue`, I can do this.

If it comes to Lombok I don't like magic around it, injecting code within the project sources (in compare to implementing interfaces as it is in Immutables) simply scares me, plus, I don't like the code it produces (bloated ifs instead of Guava oneliners). 
Finally, AFAIK it has issues with Java/JDKs changes and updates.

File: src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java (hunk 1 above)
Comment (jonaskoperdraat) [reply src/main/java/com/jarslab/maven/babel/plugin/BabelMojo.java:102]: Sure, no problem. I'll see if I can find some time tonight to work on #12

File: src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java (hunk 2 above)
Comment (jonaskoperdraat) [reply src/main/java/com/jarslab/maven/babel/plugin/BabelTranspiler.java:25]: Definately something for a separate PR

**Pull request merged**